bundles/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/se/connection/ValloxIpConnector.java

```diff
@@ -0,0 +1,144 @@
+    private final Thread readerThread = new TelegramReader("Vallox Telegram Reader");
+    private Socket socket = new Socket();
+
+    public ValloxIpConnector() {
+        logger.debug("Ip connector initialized");
+    }
+
+    /**
+     * Connect to socket
+     */
+    @Override
+    public void connect(ValloxSEConfiguration config) throws IOException {
+        if (isConnected()) {
+            return;
+        }
+        buffer.clear();
+        panelNumber = config.getPanelAsByte();
+        socket = new Socket();
+        socket.connect(new InetSocketAddress(config.tcpHost, config.tcpPort), ValloxSEConstants.CONNECTION_TIMEOUT);
+        socket.setSoTimeout(0); // Don't fail if there's nothing to read. Machine is powered down.
+        inputStream = socket.getInputStream();
+        outputStream = socket.getOutputStream();
+        logger.debug("Connected to {}:{}", config.tcpHost, config.tcpPort);
+
+        startTelegramReaderThread();
+        startProcessorThreads();
+        connected.set(true);
+    }
+
+    /**
+     * Start threads for receiving and processing telegrams
+     */
+    private void startTelegramReaderThread() {
+        if (!readerThread.isAlive()) {
+            readerThread.setDaemon(true);
+            readerThread.start();
+        }
+    }
+
+    /**
+     * Start threads for receiving and processing telegrams
+     */
+    private void stopTelegramReaderThread() {
+        readerThread.interrupt();
+        try {
+            readerThread.join(2000);
+        } catch (InterruptedException e) {
+            // Do nothing
+        }
+    }
+
+    /**
+     * Close socket and stop threads.
+     */
+    @Override
+    public void close() {
+        super.close();
+        connected.set(false);
+        logger.debug("Stopping threads and closing socket");
+        stopTelegramReaderThread();
+        stopProcessorThreads();
+        try {
+            socket.close();
+        } catch (IOException e) {
+            logger.debug("Closing socket failed: ", e);
+        }
+
+        logger.debug("Ip connection closed");
+    }
+
+    /**
+     * {@link Thread} implementation for reading telegrams
+     *
+     * @author Miika Jukka - Initial contribution
+     */
+    private class TelegramReader extends Thread {
+        boolean interrupted = false;
+
+        public TelegramReader(String name) {
+            super(name);
+        }
+
+        @Override
+        public void interrupt() {
+            interrupted = true;
+            super.interrupt();
+        }
+
+        @Override
+        public void run() {
+            logger.trace("Telegram reader thread started");
+            InputStream inputStream = getInputStream();
+            while (!interrupted && inputStream != null) {
+                try {
+                    int data = inputStream.read();
+                    while (data != -1) {
+                        buffer.add((byte) inputStream.read());
+                    }
```

Comment:
I don't think it was your intent when writing this for it to loop forever if the first read != -1.

Reply:
Fixed.